### PR TITLE
Re-implemented math classes to enable raycasting and improve precision

### PR DESCRIPTION
**Pull request by Raelr**
### The PR has been...
- [x] provided a reasonable name that is not just the branch name (e.g "Added Vulkan render delegate")
- [x] linked to its related issue
- [x] assigned a reviewer from the team
- [x] labelled appropriately

### The code has been...
- [x] made mergable and free of conflicts in relation to `master` *(according to GitHub)*
- [x] tested in a packaged state using the `package` targets
- [ ] pulled to the reviewer's machine and reasonably tested

### CODE REVIEW

File: tests/utils/test_Vec2.cpp [hunk 1]
```diff
@@ -1,333 +1,436 @@
...
+UTEST(test_Vec2, TestFuzzyEquals)
 {
-    Vec2 vec0 = {2, 3};
+    Vec2 vecA = {1.f, 2.f};
 
-    Vec2 norm = vec0.Normalise();
+    Vec2 vecB = {0.9999999999999912, 2.0000000000000023f};
 
-    ASSERT_TRUE((norm.x - 0.554775417f) < std::numeric_limits<float>::epsilon());
-    ASSERT_TRUE((norm.y - 0.832163095f) < std::numeric_limits<float>::epsilon());
```
Comment (J-Mo63) [inline tests/utils/test_Vec2.cpp:322]: Again, should test alternate case on boundary

**Review by J-Mo63 — APPROVED** [review]
Looks good, mostly docs to re-add some small changes required otherwise.

File: engine/utils/Logging.h [hunk 2]
```diff
@@ -92,9 +96,12 @@ class VariantContainer
     DEFINE_VARIANT_TYPE(unsigned long data, String::FromU32(data));
     DEFINE_VARIANT_TYPE(unsigned long long data, String::FromU64(data));
     DEFINE_VARIANT_TYPE(bool data, data ? "true" : "false");
-    DEFINE_VARIANT_TYPE(const Vec2& data, "Vector2(" + data.ToString() + ")");
-    DEFINE_VARIANT_TYPE(const Vec3& data, "Vector3(" + data.ToString() + ")");
-    DEFINE_VARIANT_TYPE(const Vec4& data, "Vector4(" + data.ToString() + ")");
+    DEFINE_VARIANT_TYPE(const Vec2& data, "Vector2(" + ToString(data) + ")");
+    DEFINE_VARIANT_TYPE(const Vec3& data, "Vector3(" + ToString(data) + ")");
+    DEFINE_VARIANT_TYPE(const Vec4& data, "Vector4(" + ToString(data) + ")");
+    //    DEFINE_VARIANT_TYPE(const Mat2& data, "Mat2(" + ToString(data) + ")");
```
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/utils/Logging.h]: Removed.

File: engine/render/renderer/platform/vulkan/IndexBuffer.h
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/render/renderer/platform/vulkan/IndexBuffer.h:1]: Done.

File: engine/render/renderer/renderer/Renderer2D.cpp [hunk 3]
```diff
@@ -232,12 +232,14 @@ void Renderer2D::RenderQuads(Vulkan::CommandBuffer& buffer, size_t index)
...
-                                vertexBufferOffset);
+        quadVBuffer.Copy(quadArr.Data(),
+                         sizeof(QuadVertex) * quadArr.Count(),
+                         sizeof(QuadVertex) * vertexBufferOffset);
+
+        uint64_t bindOffset = sizeof(QuadVertex) * vertexBufferOffset;
 
-        quadVertexBuffer.Bind(buffer, &vertexBufferOffset);
+        quadVBuffer.Bind(buffer, &bindOffset);
+        // quadVertexBuffer.Bind(buffer, &vertexBufferOffset);
```
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/render/renderer/renderer/Renderer2D.cpp]: Removed.

File: engine/utils/math/vec/Equality.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+#include <cmath>
+
+#include "Vec.h"
+
+namespace Siege
+{
+
+template<typename T>
+inline constexpr T ScaledEpsilon(T lhs, T rhs)
```
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/utils/math/vec/Equality.h:28]: Done

File: engine/utils/math/vec/Vec2.h [hunk 5]
```diff
@@ -9,188 +9,832 @@
...
-    static Vec2 Lerp(Vec2 origin, Vec2 destination, float time);
+    inline constexpr T& operator[](unsigned int index)
+    {
+        switch (index)
+        {
+            default:
+            case (0):
+                return x;
+            case (1):
+                return y;
```
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/utils/math/vec/Vec2.h]: Done for all classes.

File: engine/utils/math/vec/Vec4.h [hunk 6]
```diff
@@ -4,191 +4,3961 @@
...
-     * @return a new vector with the normalised vector coordinates
+     * The 4D vector subscript operator
+     * @param index the index of the requested dimension
+     * @return a reference to the requested value
      */
-    static Vec4 Normalise(const Vec4& vec);
+    inline constexpr T& operator[](unsigned int index)
+    {
+        switch (index)
+        {
```
[the remark this answers is not in the document]
Comment (Raelr) [reply engine/utils/math/vec/Vec4.h:305]: Changed

File: tests/utils/test_Mat2.cpp [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+    String result = ToString(matrix);
+
+    ASSERT_TRUE(str == result);
+}
+
+UTEST(test_Mat2, TestFuzzyEquals)
+{
+    Mat2 matrixA = {{1, 2}, {3, 4}};
+
```
[the remark this answers is not in the document]
Comment (Raelr) [reply tests/utils/test_Mat2.cpp:286]: Changed.

File: tests/utils/test_Projection.cpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                   {0, -0.00333333341, 0, 0},
+                   {0, 0, -0.00100009993, 0},
+                   {-1, -1, -0.000100010002, 1}};
+
+    Mat4 orthographic = Orthographic(0.f, 800.f, 600.f, 0.f, 0.1f, 1000.f);
+
+    ASSERT_TRUE(orthographic == expected);
+}
+
+UTEST(test_Projection, TestInverseProjectionMatrix)
```
[the remark this answers is not in the document]
Comment (Raelr) [reply tests/utils/test_Projection.cpp]: Changed.

File: tests/utils/test_Vec2.cpp (hunk 1 above)
Comment (Raelr) [reply tests/utils/test_Vec2.cpp:322]: Changed.

**Pull request merged**